### PR TITLE
Forward port CommandLineParser rewrite

### PR DESCRIPTION
**Pull request by som-snytt**
Tests to follow. (When I learn how to add tests.)

Bookmarking:
https://github.com/scala/scala/pull/6622
https://github.com/scala/scala/pull/5882

### CODE REVIEW

**Comment by smarter** [comment]
Can you add a link to the corresponding scala/scala PR or commit in the commit message ?

**Comment by nicolasstucki** [comment]
Could you add the links to the commit message (not only the PR description). It is to make sure it is in the git history.

**Comment by som-snytt** [comment]
Yes, I'll update. I just looked at this again, wondering whether to improve the quick'n'dirty code.

**Comment by som-snytt** [comment]
Added test and dependency on scala-testkit.

**Comment by som-snytt** [comment]
Probably spurious?
```
[error] scala.reflect.internal.FatalError: package object runtime does not have a decl currentMirror
```

**Comment by som-snytt** [comment]
I guess not spurious.
```
[error] error while loading package, class file '/root/.cache/coursier/v1/https/scala-webapps.epfl.ch/artifactory/central/org/scala-lang/scala-reflect/2.13.1/scala-reflect-2.13.1.jar(scala/reflect/macros/package.class)' is broken
[error] (class java.lang.RuntimeException/error reading Scala signature of package.class: Scala signature package has wrong version
[error]  expected: 5.0
[error]  found: 5.2 in package.class)
```

**Comment by som-snytt** [comment]
Removed testkit dependency, converted syntax, rebased.

File: compiler/src/dotty/tools/dotc/config/CommandLineParser.scala [hunk 1]
```diff
@@ -1,68 +1,96 @@
...
+   */
+  def tokenize(line: String, errorFn: String => Unit): List[String] =
+
+    var accum: List[String] = Nil
+
+    var pos   = 0
+    var start = 0
+    val qpos  = new ArrayBuffer[Int](16)    // positions of paired quotes
+
+    inline def cur    = if done then EOF else line.charAt(pos): Int
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: Why is that type ascription needed? Could it be moved to be the result type of cur instead?
```suggestion
    inline def cur: Int = if done then EOF else line.charAt(pos)
```

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: `charAt` is a char. The benefits of this style choice are to point out the problematic expression; it looks more cool; in result type position, it's like having eyeglasses wrapped in tape where they broke. That's why implicits, the nerdiest feature, requires it.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: ok, I guess that's fine.

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: It's a bit quirky, and it was so long ago, I had to reconstruct my thinking. On the up side, with all the talk about command line parsing support for main entry point, probably this code will be replaced anyway. It was mostly aimed at Process building.

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: Also thanks, this was to get my feet wet with the repo and process, low benefit and low risk.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/config/CommandLineParser.scala:26]: Thanks to you for porting this! And sorry for the long delay before the review, it slipped my mind.

**Pull request merged**